### ISSUE TITLE
Potential bug regarding componentDidUpdate prev/current props when refreshing from a mutation.

### ISSUE THREAD

**Issue by dvingo** [fulcrologic/fulcro#211]
I extracted this from another project and tried to cut down the code to a minimal amount.
The repo here demonstrates the bug (this was created with `lein new fulcro did-update-bug nodemo shadow-cljs`):
https://bitbucket.org/dvingo/fulcro_component_did_update_bug/overview

Succinctly: `componentDidUpdate` sees a boolean prop in prev-props and current props go from false/false directly to true/true without seeing false/true.

Relevant code is in this file:
https://bitbucket.org/dvingo/fulcro_component_did_update_bug/src/f7fac218c79f20c6c731b6300657af5612c38862/src/main/did_update_bug/ui/root.cljc?at=master&fileviewer=file-view-default

I also commented out :render-mode :keyframe.

To reproduce:
1. Run `yarn shadow-cljs watch main`
2. Open app - enter the correct answer for the multiplication problem.

If you open the console you'll see that the pair of previous props and current props in `AnswerPrompt`s `componentDidUpdate` go from `false` `false` directly to `true` `true`, never being called with `false` `true`.

Expected:
After you enter the correct answer the mutation `answer-prompt/next-question` should fire.

I found that commenting out the mutation refresh here:

https://bitbucket.org/dvingo/fulcro_component_did_update_bug/src/f7fac218c79f20c6c731b6300657af5612c38862/src/main/did_update_bug/ui/root.cljc?at=master&fileviewer=file-view-default#root.cljc-55

gets the correct false/true pair to be passed to the component, but then the `answers/progress-report` is stale.

**Comment by awkay** [comment]
You mentioned you'd like to try your hand at fixing this, so here's what I'd suggest:

- Clone Fulcro
- Import your example into a devcard, and use the emulated server for the server-side (see the load cards for an example)
- Add debug statements in primitives.cljc and trigger your problem

Here are the places that matter:

1. The factory (cljs version) https://github.com/fulcrologic/fulcro/blob/develop/src/main/fulcro/client/primitives.cljc#L973

The factory is where the fulcro stuff gets crammed into low-level react props. Note the `get-current-time`. This is a monotonically increasing integer so we know when props that we received were received. Think of it as the "frame number". Note also that the value of the "props" is the combo of the actual props *and* this time.

2. The lifecycle wrapper https://github.com/fulcrologic/fulcro/blob/develop/src/main/fulcro/client/primitives.cljc#L297 is used by the macro to wrap Fulcro's functionality around the React lifecycle. *Your* code in the defsc for that lifecycle ends up as `body`

ALl of the code in (1) and (2) is basically just shoving stuff into low-level js spots...once you understand WHY the time is needed, you'll understand most of it.

Imagine you have this UI Tree (all components have idents except root)

```
C1  C2
    P
  Root
```

On initial render, you have props at t=1 for all components.

You mutate C1, and it gets props for t=2. Note that P still has props at t=1 (which *include* the old props for C1 at t=1, since they were originally passed *through* it), and P has *not* re-rendered.

Now imagine that you do a react-set-state! on P.  That does *not* run a Fulcro query...it just force-updates P with the new state...but when render is called on P *it* passes *old* props to C1 and C2.  So, the `shouldComponentUpdate` wrapper on https://github.com/fulcrologic/fulcro/blob/develop/src/main/fulcro/client/primitives.cljc#L360 has logic to short-circuit the rendering of C1 or C2 if they have props that are the same or newer than what the parent is passing in. This prevents C1 from "going back in time".

However, all of this logic requires that we are careful with things like prev-props...so, that is where the bug is.

Ask any questions you have.

**Comment by dvingo** [comment]
Thanks for the details Tony! I'll dive in and let you know if I have any questions.

**Comment by dvingo** [comment]
I tracked down the problem to `prim/-prev-props` + `prim/merge-pending-props!` using the most recent (with regards to basis-t) props due to this call: `(props* <props-from-react-props> <react-state "fulcro$prev$value">)` returning `<props-from-react-props>` when a parent has rendered from above. 

When the bug appears `<props-from-react-props>` basis-t will equal the current reconciler time and `<react-state "fulcro$prev$value">` will have `basis-t == current-reconciler-t - 2`. 
So I naively check for these conditions and conditionally alter the return value based on that logic.

This feels a bit like a hack so I wanted to bring it up to get some feedback on a better approach.

It seems like the general heuristic for determining the correct previous props is to use the current implementation unless:
  - fulcro-props from react props has `basis-t == current reconciler t`
    and 
    fulro$prev$value in react state has `basis-t > 0`

 This description is probably a bit shit, so here is the diff:

https://github.com/fulcrologic/fulcro/compare/develop...dvingo:component-did-update-bug?expand=1

**Comment by awkay** [comment]
Hm, yeah that does look a bit too much like a hack.  I think the "2" is definitely not of interest...we're really just interested in the age of the props.  I need to get into this part of the code (which was largely inherited) and clean it up.  Especially since React 16 already defines lifecycle methods we don't support, and 17 is coming with async rendering.  I'm going to work on all of that today.

**Comment by awkay** [comment]
I'm reworking the React internals while I'm at it. There are a number of things that React 16.4+, in prep for 17, has changed that require fixes in Fulcro.

**Comment by awkay** [comment]
I've confirmed and reproduced this issue, along with some other lifecycle bugs. I've got a much-simplified devcard to play with it, and hope to have a fix soon.

**Comment by dvingo** [comment]
Ah great, glad to hear it. hah yea that devcard was a bit wordy, thanks for paring it down.

**Comment by awkay** [comment]
Fixed in 2.6.0  (awaiting more testing to close)
